db/query.go

```diff
@@ -0,0 +1,419 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package db
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	"firebase.google.com/go/internal"
+
+	"golang.org/x/net/context"
+)
+
+// Query represents a complex query that can be executed on a Ref.
+//
+// Complex queries can consist of up to 2 components: a required ordering constraint, and an
+// optional filtering constraint. At the server, data is first sorted according to the given
+// ordering constraint (e.g. order by child). Then the filtering constraint (e.g. limit, range) is
+// applied on the sorted data to produce the final result. Despite the ordering constraint, the
+// final result is returned by the server as an unordered collection. Therefore the values read
+// from a Query instance are not ordered.
+type Query struct {
+	client              *Client
+	path                string
+	ob                  orderBy
+	limFirst, limLast   int
+	start, end, equalTo interface{}
+}
+
+// StartAt returns a shallow copy of the Query with v set as a lower bound of a range query.
+//
+// The resulting Query will only return child nodes with a value greater than or equal to v.
+func (q *Query) StartAt(v interface{}) *Query {
+	q2 := new(Query)
```

Comment:
Here and below, &Query{} would be more consistent with our current SDK.

Reply:
Done

Reply:
Done